[SYCL][Doc] Add free function kernel extension spec (#11861)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

The return value tells the number of formal parameters in the function's
definition.

=== Enqueuing a free function kernel and setting parameter values

Once the application obtains a `kernel` object for a free function kernel, it
can enqueue the kernel to a device using any of the SYCL functions that allow
a kernel to be enqueued via a `kernel` object.
The application must enqueue the free function kernel according to its type.
For example, a free function kernel defined via `SYCL_EXT_ONEAPI_RANGE_KERNEL`
can be enqueued by calling the `handler::parallel_for` overload taking a
`range`.
A free function kernel defined via `SYCL_EXT_ONEAPI_ND_RANGE_KERNEL` can be
enqueued by calling the `handler::parallel_for` overload taking an `nd_range`.
A free function kernel defined via `SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL` can be
enqueued by calling `handler::single_task`.

Attempting to enqueue a free function kernel using a mechanism that does not
match its type results in undefined behavior.
Attempting to enqueue a free function kernel with a `range` or `nd_range` whose
dimensionality does not match the free function kernel definition results in

> undefined behavior.

Author: I was thinking that by having three different macros we can provide better error messages.  For example, we can emit a compile-time error if a kernel decorated with `SYCL_EXT_ONEAPI_RANGE_KERNEL` calls the function `sycl::ext::oneapi::this_kernel::get_nd_item`.  We can also throw an exception if a kernel decorated with `SYCL_EXT_ONEAPI_RANGE_KERNEL` is enqueued as an nd-range kernel.

Note that we can diagnose these errors even if the spec says the behavior is UB.  We always have the option to diagnose UB with a helpful error.

> undefined behavior.

Reviewer: > For example, we can emit a compile-time error if a kernel decorated with `SYCL_EXT_ONEAPI_RANGE_KERNEL` calls the function `sycl::ext::oneapi::this_kernel::get_nd_item`.

Is this a situation where we need to emit an error? I would expect any SYCL kernel to work for any combination of kernel invocation API and kernel identifiers free functions. nd_item is a wrapper around kernel identifiers, which are available to any kernel regardless of parallel_for overload. Current SYCL API limitations on lambda call operator parameter type are artificial. They are useful for code optimization: compiler has to allocate and initialize id/item/nd_item lambda argument. In this case, it's created by user via free function, so we don't need to optimize it.

> undefined behavior.

Author: I do think an error would be helpful.  The specification for the free function queries says their behavior is undefined if they are called from a kernel that is launched in an incompatible way:

> It is the user’s responsibility to ensure that these functions are called in a manner that is compatible with the kernel’s launch parameters, as detailed in the definition of each function. Calling these functions from an incompatible kernel results in undefined behavior.

(https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_queries.asciidoc#accessing-instances-of-special-sycl-classes)

If we don't say it is UB, then we would need to clarify the spec to explain what is returned in the mismatched cases.  For example, what is the work-group size for a simple "range" parallel-for?  This seems worse than saying the behavior is UB and diagnosing an error when possible.

> undefined behavior.

Reviewer: > If we don't say it is UB, then we would need to clarify the spec to explain what is returned in the mismatched cases.

There are no mismatches cases because there is no SYCL kernel argument we have to "match" (or to be compatible with). This argument exists only for call operator of functional object, but this extension defines alternative API, which we don't need to "match".

> For example, what is the work-group size for a simple "range" parallel-for?

I expect it to be defined by the implementation. AFAIK, all supported backends provide access to work-group size on device side.
On the host side, when not set explicitly OpenCL back-end chooses the work-group size automatically. Other backends has API to query recommended work-group size, which can be set explicitly.

> undefined behavior.

Author: > There are no mismatches cases because there is no SYCL kernel argument we have to "match"

By "mismatched" I meant a mismatch between the way the kernel is launched and the queries the kernel makes.  For example, launching the kernel as a simple "range" kernel and calling `this_kernel::get_nd_item`.

> I expect it to be defined by the implementation. AFAIK, all supported backends provide access to work-group size on device side.

I really feel this would open a can of worms that would make our life difficult in the future.  For example, what if the application launches the kernel as a simple "range" and then uses `this_kernel::get_nd_item` to get a `group` object that it uses to call one of the group functions defined in [section 4.17.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:group-functions) or [4.17.3](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:algorithms)?  I really don't think we want to say this is supported.  If we were required to support this, I think it would limit our implementation choices in the future.

I also think this presents a confusing story to our users.  @Pennycook has been clarifying the SYCL spec to say that simple "range" kernels have no concept of work-group.  Therefore, it does not make sense to allow "range" kernels to call `this_kernel::get_nd_item`.

`handler::single_task`, the kernel argument values must be set via
`handler::set_arg` or `handler::set_args`.
Failing to set the value of a kernel argument results in undefined behavior.
The type of the value passed to `handler::set_arg` or `handler::set_args` must
be the same as the type of the corresponding formal parameter in the free
function kernel.
Passing a value with a mismatched type results in undefined behavior.

=== Obtaining the iteration id for a kernel

In a standard SYCL kernel, the iteration ID is passed as a parameter to the
kernel's callable object.
However, this is not the case for a free function kernel because the function
parameters are used to pass the kernel arguments instead.
Therefore, a free function kernel must obtain the iteration ID in some other
way.
Typically, a free function kernel uses the functions specified in
link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
sycl_ext_oneapi_free_function_queries] for this purpose.

=== Address space of kernel arguments

The arguments to a free function kernel are in the private address space.
As a result, a kernel can modify its arguments, but the modification is visible

> only within the work-item.

Author: Yes, I agree.  This paragraph only talks about the arguments themselves, not what the pointers point to.  Are you saying that this should be made clearer?

> only within the work-item.

Reviewer: Probably, I misinterpreted the goal of this paragraph. I thought it describes all kernel arguments limitations regarding address spaces.

> only within the work-item.

Author: I clarified this in 3adda07be441a8ec1751e8905aad8be679426825.

  function kernel to be enqueued using the APIs defined in
  link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions], but we haven't settled on the exact API
  yet.
  One option is like this:
+
```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) { /*...*/  }

int main() {
  sycl::queue q;
  float *ptr = sycl::malloc_shared<float>(N, q);
  sycl::parallel_for<iota>(q, {N}, 1.f, ptr);
}
```
+
Another option is like this:
+
```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) { /*...*/  }

int main() {
  sycl::queue q;
  float *ptr = sycl::malloc_shared<float>(N, q);
  sycl::parallel_for(q, {N}, kfp<iota>, 1.f, ptr);
}
```
+
Where `kfp` would have some nicer name.

* Can the front-end define a trait like this, which returns true only if the
  address is a free function kernel:
+
```
template <auto *Func>
bool is_kernel_function_v;
```
+
If the front-end can provide this, we can provide a nice diagnostic when the
user passes an invalid address to the kernel bundle functions like
`template<auto *Func> kernel_id get_kernel_id()`.

* We need to investigate whether there will be problems passing kernel

>   arguments that are "decomposed" by the front-end.

Author: I think you are saying that we should prohibit kernel arguments that require decomposition, which means the application will not be able to pass either `accessor` or `local_accessor`.  Correct?  It might be reasonable to prohibit `accessor` because applications can use USM pointers instead.  However, prohibiting `local_accessor` seems like a big limitation because this is the main way to allocate SLM in a kernel.

I suppose we could require applications to use [sycl_ext_oneapi_work_group_local](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_local.asciidoc) if they need SLM, but that seems less convenient than using `local_accessor`.

Do you think we could allow both `accessor` and `local_accessor` as kernel arguments and teach `set_arg` to know that these arguments need to be decomposed into several low-level kernel arguments?  It seems like this might not be too difficult.  I think we could prohibit more complicated cases, for example, when an `accessor` is a member variable of a struct.

>   arguments that are "decomposed" by the front-end.

Author: > Can you define how SLM should be represented in "raw kernel"?

This depends on the outcome of this conversation.

If we agree that it is OK to pass `local_accessor` as a kernel argument, then you simply define a "raw kernel" with an argument of type `local_accessor`.  The kernel can then use any of the standard member functions of `local_accessor` to access the SLM.

>   arguments that are "decomposed" by the front-end.

Author: @bader and I had a meeting, and we decided that the extension will be able to support any kernel argument type, even those that require decomposition like `accessor` and `local_accessor`.  I resolved the related issue in c0f2029a3360ea105fbb6859ae98c025a9d120dd.

----
!====

Indicates that the function is a free function kernel that is invoked with an
`nd_range` iteration space of `Dims` dimensions.
The `Dims` parameter must be an integral constant expression that specifies a
valid number of dimensions for the `nd_range` class.

a|
[frame=all,grid=none]
!====
a!
[source]
----
#define SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL /*unspecified */
----
!====

Indicates that the function is a free function kernel that is invoked via
`single_task` (i.e. without any iteration space).
|====

When a function is defined as a free function kernel, each parameter to the
function is a kernel argument, which must abide by the rules for allowable
kernel parameter types specified in section 4.12.4 "Rules for parameter passing
to kernels" of the core SYCL specification.
The function's return type must be `void`.

The following example demonstrates how a free function kernel using a
3-dimensional nd-range iteration space can be defined:

```
SYCL_EXT_ONEAPI_ND_RANGE_KERNEL(3)
void iota(float start, float *ptr) {
   // ...
}
```

A function decorated with one of these macros can still be called from host
code.

> The macro has no effect in such cases.

Author: Yes, that's a good point.  I'll update the spec to allow this.

BTW, this relates to a question @AlexeySachkov raised.  In SPIR-V, it's illegal to call to a kernel function from anywhere else in the SPIR-V module.  We can still allow kernel functions to be called at the SYCL level, though.  It just means that the compiler needs to generate a wrapper around the kernel function.  That way the wrapper become the kernel function  at the SPIR-V level, and there will be no calls to this wrapper from the application.

I believe our device compiler already generates a wrapper like this, so I think this will fall out for free.

> The macro has no effect in such cases.

Author: I fixed this in 73ddbad24e765ad1db7d136d532818b8df55a80a.

  // to disambiguate.
  sycl::kernel_id ping_float = syclex::get_kernel_id<(void(*)(float))ping>();
  sycl::kernel_id ping_int = syclex::get_kernel_id<(void(*)(int))ping>();
}
```

== Issues

* We're pretty sure that we want to define some syntax that allows a free
  function kernel to be enqueued using the APIs defined in
  link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
  sycl_ext_oneapi_enqueue_functions], but we haven't settled on the exact API
  yet.
  One option is like this:
+
```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) { /*...*/  }

int main() {
  sycl::queue q;
  float *ptr = sycl::malloc_shared<float>(N, q);
  sycl::parallel_for<iota>(q, {N}, 1.f, ptr);
}
```
+
Another option is like this:
+
```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) { /*...*/  }

int main() {
  sycl::queue q;
  float *ptr = sycl::malloc_shared<float>(N, q);
  sycl::parallel_for(q, {N}, kfp<iota>, 1.f, ptr);
}
```
+
Where `kfp` would have some nicer name.

* Can the front-end define a trait like this, which returns true only if the
  address is a free function kernel:
+
```
template <auto *Func>
bool is_kernel_function_v;
```

> +

Reviewer 2: I would expect the trait to have to be backed by some kind of compiler magic since, as far as I know, there is no way in standard C++ to query whether a function was declared with a particular attribute. If the trait is used in host code, that would exclude support for SYCL-unaware host compilers. Presumably, the macros would expand to nothing for such host compilers anyway. Providing an implementation for device-side compilation should not pose any difficulty though.

Author: The main use case I have in mind requires this trait to be available in the host compilation pass, so it is only useful if we can make this work.

I think it does require compiler magic, so it probably cannot be done with a third-party host compiler.  The trait would still be useful if it only works when our clang is used for the host compiler pass, though.  This just means we will get better error messages when people use our clang as the host compiler.

It's *almost* possible to implement this trait in the integration header, but I can't think how to make it work in all cases.  For the cases where the kernel function is defined at namespace scope, the integration header can look like this:

```
#include <type_traits>

void foo();  // Compiler generated forward declaration of application's kernel function

// Base trait definition
template<auto *Func> struct is_kernel_function : std::false_type {};

// Specializations for each of the application's kernel functions
template<> struct is_kernel_function<foo> : std::true_type {};
```

However, I can't think to to make this work for the case when the kernel is defined as a static member function of a class.  I also can't think how to make it work if the kernel takes a parameter of some user-defined type that we can't forward declare in the integration header.  Therefore, I think we would need compiler magic to solve this.

Author: I added the traits to the spec in ae6f7ede568821463bed0744594db5a442289f1d.  I also defined special restrictions for implementations that use an integration header in 0e1be43b440e19378bddf6c9c09cfc5c1160440f.

sycl_ext_oneapi_kernel_compiler extension, and the
sycl_ext_oneapi_free_function_kernels extension provides that mechanism.
The "free function kernel" feature is a separate extension, though, because it
can also be used independently from sycl_ext_oneapi_kernel_compiler.
For example, some users may find it more familiar to define kernels as plain
functions or they might have other reasons to prefer a clean separation between
host and device code.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.
An implementation supporting this extension must predefine the macro
`SYCL_EXT_ONEAPI_FREE_FUNCTION_KERNELS`
to one of the values defined in the table below.
Applications can test for the existence of this macro to determine if the
implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Defining a free function kernel

> A free function kernel is a normal C++ function definition, where the function

Reviewer 3: Just noting here that it opens up a possibility to have `constexpr` or `inline` kernel now. But I don't think that any of those should cause problems or complicate the implementation.

Author: Definition a kernel as `inline` is possibly interesting because it would allow the same kernel to be defined multiple times (in different translation units).  The linker will discard all but one definition, so all references to the kernel will get the same address.

I also can't think of any problem this will cause the implementation, but we should include a test for this case.

declaration is decorated with one of the following macros:
`SYCL_EXT_ONEAPI_RANGE_KERNEL`, `SYCL_EXT_ONEAPI_ND_RANGE_KERNEL`, or
`SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL`.

When a function declaration is decorated with one of these macros, the

> following rules must be observed:

Reviewer 3: I assume that the same rules apply if a function _definition_ is decorated with one of those macro, right? Perhaps it is worth clarifying?

Author: My understanding is that every function definition is also a declaration of that function.  Therefore, any rules that apply to the function declaration also apply to the definition.

The core SYCL spec is also worded this way.  For example, [section 5.8](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device.attributes) "Attributes for device code" uses only the term "declaration" (not definition) when talking about the rules for the attributes).

Reviewer 2: Within the C++ standard, a definition is also a declaration. In scenarios where the distinction matters, it is necessary to be explicit ("a declaration that is not a definition" or "a definition").

> * The function must be declared at either namespace scope are in class scope as
>   a static member function.

Reviewer 2: ```suggestion
* The function must be declared at either namespace scope or at class scope as
  a static member function.
```
In order to support integration headers, I think the function must also be restricted to having a return type and parameter types that are forward declarable at namespace scope (e.g., they can't be member class types or enumeration types without an explicit underlying type).

Author: Thanks for catching the typo.

Yes, I agree that this isn't sufficient to support integration headers.  I'm working on a separate list of restrictions for that.  My thought is to write the main part of the specification assuming an implementation that does not rely on integration headers.  (E.g. it might be a single-pass implementation or an implementation that does integration at IR level.)  Then, I will have a separate section that lists additional limitations for implementation that use integration headers.

The restriction I added here (7d9b6f60bd1cf707ace7b08510a3c90cd74e6166) is a general requirement that applies to all implementations.

`single_task` (i.e. without any iteration space).
|====

When a function is defined as a free function kernel, each parameter to the
function is a kernel argument, which must abide by the rules for allowable
kernel parameter types specified in section 4.12.4 "Rules for parameter passing
to kernels" of the core SYCL specification.
The function's return type must be `void`.

The following example demonstrates how a free function kernel using a
3-dimensional nd-range iteration space can be defined:

```
SYCL_EXT_ONEAPI_ND_RANGE_KERNEL(3)
void iota(float start, float *ptr) {
   // ...
}
```

A function decorated with one of these macros can still be called from host
code.
The macro has no effect in such cases.

=== New kernel bundle member functions

This extension adds the following new functions which add kernel bundle support
for free function kernels:

|====
a|
[frame=all,grid=none]
!====
a!
[source]
----
namespace sycl::ext::oneapi::experimental {

template <auto *Func>
kernel_id get_kernel_id();

} // namespace sycl::ext::oneapi::experimental
----
!====

_Preconditions_: The address `Func` must be the address of some free function

> kernel that is defined in the calling application.

Reviewer 3: Should we specify which exception is thrown when `Func` is not an address of a free function kernel?

The same applies to other APIs defined in the extension. I see that it is documented for `ext_oneapi_get_kernel` - we probably just need to copy that "Throws" paragraph everywhere else

Author: I intentionally did not specify a "Throws" paragraph on these functions.  The current wording of "Precondition" simply means it is UB if the application passes a `Func` that is not a kernel.

I left this open until we resolve the issue "Can the front-end define a trait like [`is_kernel_function_v`]".  If we can define that trait, then I will change this paragraph from "Precondition" to "Constraint", which means that the compiler must diagnose a compile-time error when `Func` is not the address of a kernel.

In fact, this is the direction I'm leaning.  For an implementation that uses an integration header, I think the header can define the trait.  (There will be some additional restrictions on the kernel function in that case.  See my [comment above](https://github.com/intel/llvm/pull/11861#discussion_r1422522764).)

Author: This is addressed in ae6f7ede568821463bed0744594db5a442289f1d.  Now that we have a trait `is_kernel`, we can use that trait as a constraint on these functions.  Thus, the user will get a compile-time error if `Func` is not a kernel function.

parameters are used to pass the kernel arguments instead.
Therefore, a free function kernel must obtain the iteration ID in some other
way.
Typically, a free function kernel uses the functions specified in
link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
sycl_ext_oneapi_free_function_queries] for this purpose.

=== Address space of kernel arguments

The arguments to a free function kernel are in the private address space.
As a result, a kernel can modify its arguments, but the modification is visible
only within the work-item.

[_Note:_ This applies only to the arguments themselves, not to memory that the
arguments point to.
For example, with a USM pointer argument, the pointer argument itself is in the
private address space, but the memory it points to is in the global address
space.
_{endnote}_]

=== Interaction with kernel properties

If the implementation supports
link:../experimental/sycl_ext_oneapi_kernel_properties.asciidoc[
sycl_ext_oneapi_kernel_properties], a free function kernel may be decorated
with these properties by applying the properties to the function declaration as
illustrated below.

```
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)

> SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclex::work_group_size<32>)

Reviewer 4: I still find this quite dissatisfying.  I'm worried that if the `FUNCTION_PROPERTY` mechanism doesn't work for this case, it isn't actually going to work for any of the other cases. #10540 is pointing to sycl_ext_oneapi_free_function_kernels as a precedent to justify introducing yet another macro instead of using the property mechanism.

Could we add an issue saying something about how we'd ideally like to unify the mechanism used to declare all of these special behaviors and how they apply to functions?

Relevant discussion: https://github.com/intel/llvm/pull/10540#discussion_r1417519332

Author: I added an issue in 0fe104f983c9602f638681c2e08b377f99470676.

However, I'd like to come to a final decision on this soon because it will affect all code using this extension.

Author: We decided to go with the unified behavior.  We no longer need the special macros like `SYCL_EXT_ONEAPI_RANGE_KERNEL`, and we can use compile-time properties instead.  I made this change in aa94401c618610c6c8e375ec94540ec4a1951aa8.

+
```
void foo(int) {/*...*/}

void caller() {
  auto *pf = foo;  // This is ambiguous because foo(float) is forward declared
                   // in the integration header
}

SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void foo(float) {/*...*/}
```

The allowed types are:

* A {cpp} fundamental type.
* A class or struct that is defined at namespace scope.
* A scoped enumeration that is defined at namespace scope.
* An unscoped enumeration that has an explicit underlying type, where the
  enumeration is defined at namespace scope.
* A type alias to one of the above types.

== Example

=== Basic invocation

The following example demonstrates how to define a free function kernel and then
enqueue it on a device.

```
namespace syclex = sycl::ext::oneapi::experimental;

static constexpr size_t NUM = 1024;

SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, float *ptr) {
  // Get the ID of this kernel iteration.
  size_t id = syclex::this_kernel::get_id();

  ptr[id] = start + static_cast<float>(id);
}

void main() {
  sycl::queue q;
  sycl::context ctxt = q.get_context();

  // Get a kernel bundle that contains the free function kernel "iota".
  auto exe_bndl =

>     syclex::get_kernel_bundle<iota, sycl::bundle_state::executable>(ctxt);

Reviewer 5: Is there any way to force creation of kernel bundle from SPIR-V or device binary?
When you do i.e. AOT compilation, possibly you have these 2 types of binaries available. Forcing use of SPIR-V can be useful as debug method

Author: This is what the second template parameter controls.  In this example, the parameter is `sycl::bundle_state::executable`, which means the bundle will contain native code.  If the ELF file does not contain native code (i.e. it was not AOT compiled), then the SYCL RT will online compile the kernel from SPIR-V.

If the application wants a kernel bundle with SPIR-V, it can do so by using `sycl::bundle_state::input` for the template parameter like this:

```
auto exe_bndl =
    syclex::get_kernel_bundle<iota, sycl::bundle_state::input>(ctxt);
```

Note, however, that there is no guarantee that the ELF file has a SPIR-V representation for every kernel.  Therefore, your application should check to see if the bundle really does contain the kernel.  See this statement from the SYCL spec:

> Since the implementation may not represent all kernels in `bundle_state::input` or `bundle_state::object`, calling this function with one of those states may return a bundle that is missing some of the kernels in `kernelIds`. The application can test for this via `kernel_bundle::has_kernel()`.

(https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_obtaining_a_kernel_bundle)

The current DPC++ behavior when AOT compiling is to *not* store the a SPIR-V copy of the kernel in the ELF file unless you specifically ask for both AOT and SPIR-V.  However, we are thinking of switching the default, so that AOT *does* store a SPIR-V copy by default (unless you specifically ask for it to be omitted).

P.S.: Note that none of this is specific to the extension proposed in this PR.  Everything I said above applies also to regular SYCL kernels.

Author: I think this issue is resolved, so I'm marking it "resolved".  @bsochack, if you have more concerns, please feel free to reopen it.

  // Get a kernel object for the "iota" function from that bundle.
  sycl::kernel k_iota = exe_bndl.ext_oneapi_get_kernel<iota>();

  float *ptr = sycl::malloc_shared<float>(NUM, q);
  q.submit([&](sycl::handler &cgh) {
    // Set the values of the kernel arguments.
    cgh.set_args(3.14f, ptr);

    cgh.parallel_for({NUM}, k_iota);
  }).wait();
}
```

=== Free function kernels which are templates or overloaded

A free function kernel may be defined as a function template.
It is also legal to define several overloads for a free function kernel.
The following example demonstrates how to get a kernel identifier in such
cases.

```
namespace syclex = sycl::ext::oneapi::experimental;

template<typename T>
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(T start, T *ptr) {
  // ...
}

SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
void ping(float *x) {
  // ...
}

SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
void ping(int *x) {
  // ...
}

int main() {
  // When the free function kernel is templated, pass the address of a

>   // specific instantiation.

Reviewer 6: When the free function kernel is templated, which dtypes are used to instantiate it, since the host code that instantiates it may not be compiled with the kernel code? 

Also, if there are multiple template types (as in the example below), are all allowed combinations of these template types going to be instantiated during the compilation?

`template<typename T, typename V>
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(T start, V *ptr) {
  // ...
}`

Author: > When the free function kernel is templated, which dtypes are used to instantiate it, since the host code that instantiates it may not be compiled with the kernel code?

This PR only adds the ability to define a "free function kernel" in a SYCL application, where both host and device code are mixed together in the same source file.  Therefore, the situation you are asking about does not exist.

Are you asking about a case where the kernel is defined as a string and then compiled with the kernel compiler (i.e. "online compiled")?  This is covered in a different PR (#11985).  In particular, see the section titled ["Instantiating templated kernel functions"](https://github.com/intel/llvm/blob/21580c6006012891f1c997bb04fb1c382c5b85fc/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#instantiating-templated-kernel-functions) in that PR.

> Also, if there are multiple template types (as in the example below), are all allowed combinations of these template types going to be instantiated during the compilation?

No, the compiler only instantiates the template with the arguments that you provide.  When the free function kernel is defined in a SYCL application (i.e. not as a string that is online compiled), the application provides the template parameters in the call to `get_kernel_id` as shown in the lines that follow this comment.  Note that the expression `iota<float>` is a normal C++ expression for the address of a a templated function.  By normal C++ rules, this causes the template to be instantiated with the `float` type, so there is no magic SYCL logic here about instantiating free function kernels that are templated.  We're just relying on the normal C++ behavior.

When the kernel is defined as a string, the template is only instantiated with the template parameters that the application provides via the `registered_kernel_names` property.  Again, see the section ["Instantiating templated kernel functions"](https://github.com/intel/llvm/blob/21580c6006012891f1c997bb04fb1c382c5b85fc/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#instantiating-templated-kernel-functions) for details.

Reviewer 6: Thanks for the explanation. I wasn't referring to the case of kernel defined as a string which is online compiled. This means, when the free function kernel is provided in a SYCL source file, the template function must be instantiated from the same source file as shown in the example below.

Author: Normal C++ rules require a template to be defined in the same translation unit (i.e. source file) before the template can be instantiated.  If a template needs to be shared across many source files, the normal practice is to define the template in a header file that is included by the different source files.  If the template is instantiated the same way in two or more source files, normal C++ rules will merge them together, effectively choosing just one of the instantiations in the final executable.

None of the behavior about this is specific to SYCL.  It's just the normal rules that apply to C++ templates.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.
OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 8 specification.
All references below to the "core SYCL specification" or to section numbers in
the SYCL specification refer to that revision.

This extension also depends on the following other SYCL extensions:

* link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
  sycl_ext_oneapi_free_function_queries]

== Status

This is a proposed extension specification, intended to gather community
feedback.
Interfaces defined in this specification may not be implemented yet or may be
in a preliminary state.
The specification itself may also change in incompatible ways before it is
finalized.
*Shipping software products should not rely on APIs defined in this
specification.*

== Overview

This extension introduces a new way to define a kernel as a simple C++

> function, where the kernel arguments are parameters to the function.

Author: Resolving this because we decided *not* to use the term "C++ free function" (see comments below).

sycl_ext_oneapi_kernel_compiler extension, and the
sycl_ext_oneapi_free_function_kernels extension provides that mechanism.
The "free function kernel" feature is a separate extension, though, because it
can also be used independently from sycl_ext_oneapi_kernel_compiler.
For example, some users may find it more familiar to define kernels as plain
functions or they might have other reasons to prefer a clean separation between
host and device code.

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.
An implementation supporting this extension must predefine the macro
`SYCL_EXT_ONEAPI_FREE_FUNCTION_KERNELS`
to one of the values defined in the table below.
Applications can test for the existence of this macro to determine if the
implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|The APIs of this experimental extension are not versioned, so the
 feature-test macro always has this value.
|===

=== Defining a free function kernel

> A free function kernel is a normal C++ function definition, where the function

Author: This comment made me wonder whether the term "free function" is really want we want to use here.  I'm not sure there is a strict definition for this term.  As far as I can tell, it's not a term from the C++ specification.

This extension allows the function to be defined either at namespace scope or at class scope as a static member function.  (This is stated in the first bullet item below.)

I *think* people usually use the term "free function" to refer to a function that is not a member function.  In that case, "C++ free function" is not really the right term because we also allow the function to be a static member function.  Therefore, I think we should not change the spec to say the function must be a "C++ free function".

This also makes me wonder whether we should rename the extension to something like "plain function kernels".  I don't know of a formal concise term that means "a function declared at namespace scope or a static member function".  We could also just keep the current extension name.  I think the text of the extension clearly states which functions are allowed.

> A free function kernel is a normal C++ function definition, where the function

Reviewer 7: Yeah I see what you mean, "free function" might not be the best term for this. I think "free function" is the colloquial term that's used, but the C++ specification seems to use the term "non-member function".

So yeah this extension is supporting non-member functions and static member functions (as well as non-member function templates and static member function templates), I don't think there is a collective term for this.

I think "plain function kernels" is a good alternative, but it maybe excludes templates, maybe it's better to just to name them "function kernels"?

> A free function kernel is a normal C++ function definition, where the function

Author: We had a phone conversation about this and decided that the term "function kernel" is also confusing because the core SYCL spec uses the term "kernel function" to refer to the kernels that are defined as a lambda expression or named kernel object.  Therefore, we decided to just continue with the current name "free function kernel".

declaration is decorated with one of the following compile-time properties:

> `range_kernel`, `nd_range_kernel`, or `single_task_kernel`.

Reviewer 7: Yeah that makes sense, thanks.

> When a function declaration is decorated with one of these properties, the

Reviewer 7: Ah okay, I missed that this is only highlighting restrictions, that makes sense.

following rules must be observed:

* The function must be declared at either namespace scope or at class scope as
  a static member function.

* The function's return type must be `void`.

* The function must not accept variadic arguments.

* Each of the function's arguments must have a type that is an allowed kernel
  parameter type as specified in section 4.12.4 "Rules for parameter passing to
  kernels" of the core SYCL specification.
  The function must not be declared with parameters of type `reducer` or
  `kernel_handler`.
  These special kernel arguments cannot be passed to a free function kernel.

* All declarations of the function must provide no default parameter values.

> * The property must appear on the first declaration of the function in the

Reviewer 7: Since these properties behave much like C++ attributes I would lean towards saying there should be a compile-time error in this case, as C++ compilers typically produce an error for the equivalent misuse of attributes. Whether we require this in the specification I think could be debated, though I would probably say it should, as the alternative would be much harder for users to diagnose.

> * The property must appear on the first declaration of the function in the

Author: For now, I've captured this under "Implementation notes".  I agree that DPC++ should diagnose this error.  I also added an open issue about requiring these diagnostics for any implementation of this extension.  See 4d4535332b85e2e9ff88828b5f8baf9bd16a4a0d and 6ed0ec5e4f889c7c6b92f616313437e293a18ab9.

----
!====

Indicates that the function is a free function kernel that is invoked with an
`nd_range` iteration space of `Dims` dimensions.

The `property_value` struct has the following member variables:

[%header,cols="1,1"]
!====
!Member
!Description

a!
[source]
----
static constexpr int dimensions = Dims
----
!
The number of dimensions of the kernel's range.
!====

a|
*`single_task_kernel` property*

[frame=all,grid=none]
!====
a!
[source]
----
namespace sycl::ext::oneapi::experimental {

struct single_task_kernel_key {
  using value_t = property_value<single_task_kernel_key>;
};

inline constexpr single_task_kernel_key::value_t single_task_kernel;

template<>
struct is_property_key<single_task_kernel_key> : std::true_type {};

} // namespace sycl::ext::oneapi::experimental
----
!====

Indicates that the function is a free function kernel that is invoked via
`single_task` (i.e. without any iteration space).
|====

When a function is defined as a free function kernel, each parameter to the
function is a kernel argument.

The following example demonstrates how a free function kernel using a
3-dimensional nd-range iteration space can be defined:

```

> SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::nd_range_kernel<3>))

Reviewer 7: Ah see, that makes sense, thanks.

If `Func` is the address of a function whose declaration is decorated with any
of the properties `range_kernel`, `nd_range_kernel`, or `single_task_kernel`;
the `is_kernel<Func>` trait provides the member constant `value` equal to
`true`.
Otherwise `value` is `false`.

The helper trait `is_kernel_v` provides the value of `value`.
|====

=== New kernel bundle member functions

This extension adds the following new functions which add kernel bundle support
for free function kernels.

[_Note:_ Many of the functions in this section have a template parameter
`Func`, which identifies a free function kernel.
This kernel function may be defined in any translation unit in the application.
_{endnote}_]

|====
a|
[frame=all,grid=none]
!====
a!
[source]
----
namespace sycl::ext::oneapi::experimental {

template <auto *Func>
kernel_id get_kernel_id();

} // namespace sycl::ext::oneapi::experimental
----
!====

_Constraints_: Available only if `is_kernel_v<Func>` is `true`.

_Returns:_ The kernel identifier that is associated with that kernel.

!====
a!
[source]
----
namespace sycl::ext::oneapi::experimental {

template <auto *Func, bundle_state State>                                // (1)

> kernel_bundle<State> get_kernel_bundle(const context& ctxt);

Reviewer 7: Thanks for clarifying, sorry I realise now I added this comment and then realised it's answered later on in the document, but I forgot to come back and remove it.

are probably defined in the `<sycl/sycl.hpp>` header, which does not get
get included until after the integration header.)

As a result, these implementations impose additional restrictions for functions
that are declared as free function kernels:

* The function must be declared at namespace scope.

* Any type used in the declaration of a parameter must be one of the allowed
  types listed below.

* If the function is instantiated from a template, any type used to instantiate
  the template must be one of the allowed types listed below.

* Uses of function identifiers in the application must assume that the free
  function kernels are forward declared at the top of the translation unit.
  Note that this can also affect references to functions that are not declared
  as free functions kernels as illustrated below.
+
```
void foo(int) {/*...*/}

void caller() {
  auto *pf = foo;  // This is ambiguous because foo(float) is forward declared
                   // in the integration header
}

SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
void foo(float) {/*...*/}
```

The allowed types are:

* A {cpp} fundamental type.

> * A class or struct that is defined at namespace scope.

Reviewer 5: @gmlueck can you clarify this requirement? What does it mean defined at namespace scope?
The structure/class needs to be defined in the same namespace as a free function kernel?

Author: Namespace scope is sometimes referred to as "file scope".  For example, this restriction prohibits using a struct/class type that is defined inside a function or inside another struct or class.  Here are some examples to illustrate:

```
struct foo1;  // defined at namespace scope

namespace bar {
struct foo2;  // defined at namespace scope
}

namespace {
struct foo3;  // defined at namespace scope
}

void func() {
  struct foo4;  // NOT defined at namespace scope (this is block scope)
}

struct outer {
  struct foo5;  // NOT defined at namespace scope (this is class scope)
};
```

> The structure/class needs to be defined in the same namespace as a free function kernel?

No, this is not necessary.  The struct/class can be defined in a different namespace than the function.  For example, this is OK:

```
namespace ns1 {
struct mystruct;
}

namespace ns2 {

SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
void iota(ns1::mystruct *p) { /*...*/  }

}
```

Reviewer 5: Thank you for clarification.
What if the structure is defined at the file scope in another file? i.e. shared header?
According to what you wrote, it should be fine

Author: Yes, that's fine.  There are NO limitations on how you split your code between the main *.cpp file and the header files it includes.
